### ISSUE TITLE
Maintaining atomic git history

### ISSUE THREAD

**Issue by erthalion** [falcosecurity/libs#1119]
Hi,

TL;DR The topic was discussed few weeks ago. Here is a proposal to change the
way PRs are merged into the main branch.

# State of things

Currently, PRs are merged into the main branch as is. That's it, if during the
development the PR has a history like:

    - initial commit (signed-off-by: A, broken)
    - fixes (signed-off-by: A, broken)
    - test improvements (signed-off-by: B, works)
    - final polishing (signed-off-by: A, works)

it will land in exactly this way. There are following issues with this
approach:

* It's much harder to read through the project history.
* The changes are not atomic anymore, we've got the main branch broken for two
  commits.

Both issues make it much harder to troubleshoot bugs and complicates forks
management.

# Proposal

There are two different ways how to address the problem.

## Squash commits before merge

Change the repo settings to use squash strategy to merge PRs. In this case the
example above would be transformed (the maintainer, who's going to merge the
PR, has to adjust the commit message) into a single commit:

    - Feature XYZ
      initial commit + fixes
      signed-off-by: A
      signed-off-by: B

This approach was discussed few weeks ago, and one concern raised was about the
signed-off part, whether there are any differences between the original
approach from the legal side. My understanding is that sinse the final commit
contains all the needed signed-off-by tags, they're equivalent.

## Require rebasing as part of the contribution rules

This will require stating more clear and maintain more rigorously the
contribution requirements. This is what happens in some projects, e.g. (from
top of my head) kubernetes [[1]] or timescaledb [[2]], both of which also
require signing off commits in some way.

At the end, it requires more efforts and puts all the responsibility for the
atomic history on the contributors and on the maintainers to do a gate keeping.
The final result would be equivalent to the previous example, a single commit
with needed signed-off-by directly from the author.

[1]: https://github.com/kubernetes/community/blob/master/contributors/guide/github-workflow.md
[2]: https://github.com/timescale/timescaledb/blob/abb6762450fb90aae5536641fef85cfd3c75b510/CONTRIBUTING.md?plain=1#L68

**Comment by leogr** [comment]
Hey @erthalion 

Thank you for reporting your proposals here!

The first option (squash all commits) would not clarify the history if completely different commits were merged into one. Also, the sign-off is not equivalent. Maintainers should refrain from signing-off lines submitted by other folks. Also, It would make no sense to add as a co-author to the squashed commit, a reviewer who only suggested changing one line.


On the other hand, I found the latter proposal more interesting, especially if we use [conventional commit messages](https://www.conventionalcommits.org/en/v1.0.0/) or similar. The only issue I see with that is it will make the contribution barrier higher.

Anyway, let's discuss it!
